// google-feed/lib/index.js
var Request = require('request');

/**
 * @param {String} query
 * @param {Object} options
 * @return {Promise}
 */
exports.find = (query, options) => {
  'use strict';

  return method('find', query, options);
}


/**
 * @param {String} query
 * @param {Object} options
 * @return {Promise}
 */
exports.load = (query, options) => {
  'use strict';

  return method('load', query, options);
}


/**
 * @param {String} method
 * @param {String} query
 * @param {Object} options
 * @return {Promise}
 */
function method(method, query, options) {
  'use strict';

  let queryOptions = new Object(options);
  let defaultOptions = {
    q: query,
    output: 'json',
    v: '1.0'
  };

  return new Promise((resolve, reject) => {
    Request({
      url: `https://ajax.googleapis.com/ajax/services/feed/${method}`,
      qs: Object.assign(queryOptions, defaultOptions),
      json: true
    }, function(error, response, body) {
      if (error) {
        reject(error);
      } else if (body.responseStatus !== 200) {
        reject(body.responseDetails);
      } else {
        resolve(body.responseData);
      }
    });
  });
}
